packages/network-controller/tests/NetworkController.test.ts

```diff
@@ -3616,6 +3622,204 @@ describe('NetworkController', () => {
       });
     });
   });
+
+  describe('rollbackToPreviousProvider', () => {
+    it('should overwrite the current provider with the previous provider when current provider has type "rpc" and previous provider has type "mainnet"', async () => {
+      const messenger = buildMessenger();
+      const networkConfiguration = {
+        rpcUrl: 'https://mock-rpc-url',
+        chainId: '0xtest',
+        ticker: 'TEST',
+        nickname: undefined,
+        id: 'testNetworkConfigurationId',
+      };
+
+      const initialProviderConfig = {
+        ...buildProviderConfig({
+          type: MAINNET,
+          chainId: '1',
+          ticker: 'ETH',
+        }),
+      };
+      await withController(
+        {
+          messenger,
+          state: {
+            networkConfigurations: {
+              testNetworkConfigurationId: networkConfiguration,
+            },
+            providerConfig: initialProviderConfig,
+          },
+        },
+        async ({ controller }) => {
+          const setCurrentAsPreviousSpy = jest.spyOn(
+            controller,
+            'setCurrentAsPreviousProvider',
+          );
+          const fakeMetamaskProvider = buildFakeMetamaskProvider();
+          createMetamaskProviderMock.mockReturnValue(fakeMetamaskProvider);
+          controller.setActiveNetwork('testNetworkConfigurationId');
+          expect(setCurrentAsPreviousSpy).toHaveBeenCalledTimes(1);
+          expect(controller.state.providerConfig).toStrictEqual({
+            ...networkConfiguration,
+            type: RPC,
+          });
+          controller.rollbackToPreviousProvider();
+          expect(controller.state.providerConfig).toStrictEqual(
+            initialProviderConfig,
+          );
+        },
+      );
+    });
+
+    it('should overwrite the current provider with the previous provider when current provider has type "mainnet" and previous provider has type "rpc"', async () => {
+      const messenger = buildMessenger();
+      const networkConfiguration = {
+        rpcUrl: 'https://mock-rpc-url',
+        chainId: '0xtest',
+        ticker: 'TEST',
+        nickname: undefined,
+        id: 'testNetworkConfigurationId',
+      };
+
+      const initialProviderConfig = {
+        ...buildProviderConfig({
+          ...networkConfiguration,
+        }),
+        type: RPC as NetworkType,
```

Comment:
Instead of typecasting, can we use `'rpc' as const`? (Edit: Or we convert `NetworkType` to a enum and then use `NetworkType.Rpc`.)